<h1>Testing Python</h1>

In [1]:
print("Wello Horld!")

Wello Horld!


<h1>Downloading Necessary Libraries</h1>

In [2]:
pip install numpy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
pip install pandas

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [4]:
pip install matplotlib

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
pip install seaborn

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


<h1>Importing Installed Libraries</h1>

In [6]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

<h1>Importing Dataset to Data Frame</h1>

In [7]:
birds = pd.read_csv('datasets\\penguins.csv')

In [8]:
birds

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen            39.1           18.7   
1        2     Adelie  Torgersen            39.5           17.4   
2        3     Adelie  Torgersen            40.3           18.0   
3        4     Adelie  Torgersen             NaN            NaN   
4        5     Adelie  Torgersen            36.7           19.3   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream            55.8           19.8   
340    341  Chinstrap      Dream            43.5           18.1   
341    342  Chinstrap      Dream            49.6           18.2   
342    343  Chinstrap      Dream            50.8           19.0   
343    344  Chinstrap      Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  
0                181.0       3750.0    male  2007  
1                186.0       3800.0  female  2007  
2                195.0       3250.0  female  2007  
3                  NaN          NaN     NaN  2007  
4                193.0       3450.0  female  2007  
..                 ...          ...     ...   ...  
339              207.0       4000.0    male  2009  
340              202.0       3400.0  female  2009  
341              193.0       3775.0    male  2009  
342              210.0       4100.0    male  2009  
343              198.0       3775.0  female  2009  

[344 rows x 9 columns]

<h1>Checking your Data Frame for null values, and data types</h1>

In [9]:
birds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   rowid              344 non-null    int64  
 1   species            344 non-null    object 
 2   island             344 non-null    object 
 3   bill_length_mm     342 non-null    float64
 4   bill_depth_mm      342 non-null    float64
 5   flipper_length_mm  342 non-null    float64
 6   body_mass_g        342 non-null    float64
 7   sex                333 non-null    object 
 8   year               344 non-null    int64  
dtypes: float64(4), int64(2), object(3)
memory usage: 24.3+ KB


In [10]:
birds.isnull()

rowid  species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    False    False   False           False          False              False   
1    False    False   False           False          False              False   
2    False    False   False           False          False              False   
3    False    False   False            True           True               True   
4    False    False   False           False          False              False   
..     ...      ...     ...             ...            ...                ...   
339  False    False   False           False          False              False   
340  False    False   False           False          False              False   
341  False    False   False           False          False              False   
342  False    False   False           False          False              False   
343  False    False   False           False          False              False   

     body_mass_g    sex   year  
0          False  False  False  
1          False  False  False  
2          False  False  False  
3           True   True  False  
4          False  False  False  
..           ...    ...    ...  
339        False  False  False  
340        False  False  False  
341        False  False  False  
342        False  False  False  
343        False  False  False  

[344 rows x 9 columns]

In [11]:
birds.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h1>Descriptive Analytics</h1>

In [12]:
birds.describe()

rowid  bill_length_mm  bill_depth_mm  flipper_length_mm  \
count  344.000000      342.000000     342.000000         342.000000   
mean   172.500000       43.921930      17.151170         200.915205   
std     99.448479        5.459584       1.974793          14.061714   
min      1.000000       32.100000      13.100000         172.000000   
25%     86.750000       39.225000      15.600000         190.000000   
50%    172.500000       44.450000      17.300000         197.000000   
75%    258.250000       48.500000      18.700000         213.000000   
max    344.000000       59.600000      21.500000         231.000000   

       body_mass_g         year  
count   342.000000   344.000000  
mean   4201.754386  2008.029070  
std     801.954536     0.818356  
min    2700.000000  2007.000000  
25%    3550.000000  2007.000000  
50%    4050.000000  2008.000000  
75%    4750.000000  2009.000000  
max    6300.000000  2009.000000

<h1>Transpose</h1>

In [13]:
birds.describe().T

count         mean         std     min       25%      50%  \
rowid              344.0   172.500000   99.448479     1.0    86.750   172.50   
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                       75%     max  
rowid               258.25   344.0  
bill_length_mm       48.50    59.6  
bill_depth_mm        18.70    21.5  
flipper_length_mm   213.00   231.0  
body_mass_g        4750.00  6300.0  
year               2009.00  2009.0

In [14]:
birds['body_mass_g']

0      3750.0
1      3800.0
2      3250.0
3         NaN
4      3450.0
        ...  
339    4000.0
340    3400.0
341    3775.0
342    4100.0
343    3775.0
Name: body_mass_g, Length: 344, dtype: float64

In [15]:
birds['sex']

0        male
1      female
2      female
3         NaN
4      female
        ...  
339      male
340    female
341      male
342      male
343    female
Name: sex, Length: 344, dtype: object

<h1>How to Define or Create your own column</h1>

In [16]:
birds['weight'] = np.round(birds['body_mass_g'] / 1000, 1)

In [17]:
birds

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen            39.1           18.7   
1        2     Adelie  Torgersen            39.5           17.4   
2        3     Adelie  Torgersen            40.3           18.0   
3        4     Adelie  Torgersen             NaN            NaN   
4        5     Adelie  Torgersen            36.7           19.3   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream            55.8           19.8   
340    341  Chinstrap      Dream            43.5           18.1   
341    342  Chinstrap      Dream            49.6           18.2   
342    343  Chinstrap      Dream            50.8           19.0   
343    344  Chinstrap      Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  weight  
0                181.0       3750.0    male  2007     3.8  
1                186.0       3800.0  female  2007     3.8  
2                195.0       3250.0  female  2007     3.2  
3                  NaN          NaN     NaN  2007     NaN  
4                193.0       3450.0  female  2007     3.4  
..                 ...          ...     ...   ...     ...  
339              207.0       4000.0    male  2009     4.0  
340              202.0       3400.0  female  2009     3.4  
341              193.0       3775.0    male  2009     3.8  
342              210.0       4100.0    male  2009     4.1  
343              198.0       3775.0  female  2009     3.8  

[344 rows x 10 columns]

In [18]:
birds.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
weight                2
dtype: int64

<h1>What to do with Null Values?</h1>

In [19]:
birds['body_mass_g'].mean()

np.float64(4201.754385964912)

In [20]:
birds['sex'].mode()

0    male
Name: sex, dtype: object

In [21]:
birds['bill_length_mm'] = birds['bill_length_mm'].fillna(birds['bill_length_mm'].mean())

In [22]:
birds.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        0
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
weight                2
dtype: int64

In [23]:
birds['sex'] = birds['sex'].fillna(birds['sex'].mode()[0])

In [24]:
birds.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        2
flipper_length_mm    2
body_mass_g          2
sex                  0
year                 0
weight               2
dtype: int64

In [25]:
birds['bill_depth_mm'] = birds['bill_depth_mm'].fillna(birds['bill_depth_mm'].mean())

In [26]:
birds.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    2
body_mass_g          2
sex                  0
year                 0
weight               2
dtype: int64

In [27]:
birds['flipper_length_mm'] = birds['flipper_length_mm'].fillna(birds['flipper_length_mm'].mean())

In [28]:
birds.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          2
sex                  0
year                 0
weight               2
dtype: int64

In [29]:
birds['body_mass_g'] = birds['body_mass_g'].fillna(birds['body_mass_g'].mean())

In [30]:
birds.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
weight               2
dtype: int64

In [31]:
birds['weight'] = birds['weight'].fillna(birds['weight'].mean())

In [32]:
birds.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
weight               0
dtype: int64

In [33]:
np.round(birds, 1)

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen            39.1           18.7   
1        2     Adelie  Torgersen            39.5           17.4   
2        3     Adelie  Torgersen            40.3           18.0   
3        4     Adelie  Torgersen            43.9           17.2   
4        5     Adelie  Torgersen            36.7           19.3   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream            55.8           19.8   
340    341  Chinstrap      Dream            43.5           18.1   
341    342  Chinstrap      Dream            49.6           18.2   
342    343  Chinstrap      Dream            50.8           19.0   
343    344  Chinstrap      Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  weight  
0                181.0       3750.0    male  2007     3.8  
1                186.0       3800.0  female  2007     3.8  
2                195.0       3250.0  female  2007     3.2  
3                200.9       4201.8    male  2007     4.2  
4                193.0       3450.0  female  2007     3.4  
..                 ...          ...     ...   ...     ...  
339              207.0       4000.0    male  2009     4.0  
340              202.0       3400.0  female  2009     3.4  
341              193.0       3775.0    male  2009     3.8  
342              210.0       4100.0    male  2009     4.1  
343              198.0       3775.0  female  2009     3.8  

[344 rows x 10 columns]

In [35]:
birds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   rowid              344 non-null    int64  
 1   species            344 non-null    object 
 2   island             344 non-null    object 
 3   bill_length_mm     344 non-null    float64
 4   bill_depth_mm      344 non-null    float64
 5   flipper_length_mm  344 non-null    float64
 6   body_mass_g        344 non-null    float64
 7   sex                344 non-null    object 
 8   year               344 non-null    int64  
 9   weight             344 non-null    float64
dtypes: float64(5), int64(2), object(3)
memory usage: 27.0+ KB


In [36]:
birds['year'] = birds['year'].astype('category')

In [37]:
birds['species'] = birds['species'].astype('category')

In [38]:
birds['rowid'] = birds['rowid'].astype('category')

In [39]:
birds['island'] = birds['island'].astype('category')

In [40]:
birds['sex'] = birds['sex'].astype('category')

<h1>Conditional Operators</h1>

In [41]:
# NOTES:
# == != used for qualitative
# >= <= used for quantitative

In [42]:
birds[birds['species'] == 'Adelie'].count()

rowid                152
species              152
island               152
bill_length_mm       152
bill_depth_mm        152
flipper_length_mm    152
body_mass_g          152
sex                  152
year                 152
weight               152
dtype: int64

In [43]:
birds[birds['species'] == 'Gentoo'].count()

rowid                124
species              124
island               124
bill_length_mm       124
bill_depth_mm        124
flipper_length_mm    124
body_mass_g          124
sex                  124
year                 124
weight               124
dtype: int64

In [44]:
birds[birds['species'] == 'Chinstrap'].count()

rowid                68
species              68
island               68
bill_length_mm       68
bill_depth_mm        68
flipper_length_mm    68
body_mass_g          68
sex                  68
year                 68
weight               68
dtype: int64

In [45]:
x = birds[birds['species'] == 'Adelie'].count()
y = birds[birds['species'] == 'Gentoo'].count()
z = birds[birds['species'] == 'Chinstrap'].count()

<h1>How many Gentoo species have weight more than 3kg</h1>

In [46]:
gentoo_count = 0

for a in y:
    if birds['weight'].all() <= 3:
        gentoo_count += 1
print(gentoo_count)

10


In [47]:
birds['species'] == "Gentoo"

0      False
1      False
2      False
3      False
4      False
       ...  
339    False
340    False
341    False
342    False
343    False
Name: species, Length: 344, dtype: bool

In [48]:
birds['weight'] > 3

0      True
1      True
2      True
3      True
4      True
       ... 
339    True
340    True
341    True
342    True
343    True
Name: weight, Length: 344, dtype: bool

In [49]:
birds[birds['species'] == 'Gentoo']

rowid species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152   153  Gentoo  Biscoe        46.10000       13.20000         211.000000   
153   154  Gentoo  Biscoe        50.00000       16.30000         230.000000   
154   155  Gentoo  Biscoe        48.70000       14.10000         210.000000   
155   156  Gentoo  Biscoe        50.00000       15.20000         218.000000   
156   157  Gentoo  Biscoe        47.60000       14.50000         215.000000   
..    ...     ...     ...             ...            ...                ...   
271   272  Gentoo  Biscoe        43.92193       17.15117         200.915205   
272   273  Gentoo  Biscoe        46.80000       14.30000         215.000000   
273   274  Gentoo  Biscoe        50.40000       15.70000         222.000000   
274   275  Gentoo  Biscoe        45.20000       14.80000         212.000000   
275   276  Gentoo  Biscoe        49.90000       16.10000         213.000000   

     body_mass_g     sex  year    weight  
152  4500.000000  female  2007  4.500000  
153  5700.000000    male  2007  5.700000  
154  4450.000000  female  2007  4.400000  
155  5700.000000    male  2007  5.700000  
156  5400.000000    male  2007  5.400000  
..           ...     ...   ...       ...  
271  4201.754386    male  2009  4.202047  
272  4850.000000  female  2009  4.800000  
273  5750.000000    male  2009  5.800000  
274  5200.000000  female  2009  5.200000  
275  5400.000000    male  2009  5.400000  

[124 rows x 10 columns]

In [50]:
birds['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [51]:
birds['species'] == 'Gentoo'

0      False
1      False
2      False
3      False
4      False
       ...  
339    False
340    False
341    False
342    False
343    False
Name: species, Length: 344, dtype: bool

In [52]:
birds['island'].value_counts()

island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64

In [53]:
birds[(birds['species'] == 'Gentoo') & (birds['island'].value_counts())]

C:\Users\LAB4-StudentPC38\AppData\Local\Temp\ipykernel_1648\2479856040.py:1: FutureWarning: Operation between non boolean Series with different indexes will no longer return a boolean result in a future version. Cast both Series to object type to maintain the prior behavior.
  birds[(birds['species'] == 'Gentoo') & (birds['island'].value_counts())]
C:\Users\LAB4-StudentPC38\AppData\Local\Temp\ipykernel_1648\2479856040.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  birds[(birds['species'] == 'Gentoo') & (birds['island'].value_counts())]


Empty DataFrame
Columns: [rowid, species, island, bill_length_mm, bill_depth_mm, flipper_length_mm, body_mass_g, sex, year, weight]
Index: []

<h1>Species among the islands</h1>

In [54]:
adelie_torgersen = birds[(birds['island'] == 'Torgersen') & (birds['species'] == 'Adelie')]
adelie_dream = birds[(birds['island'] == 'Dream') & (birds['species'] == 'Adelie')]
adelie_biscoe = birds[(birds['island'] == 'Biscoe') & (birds['species'] == 'Adelie')]
gentoo_torgersen = birds[(birds['island'] == 'Torgersen') & (birds['species'] == 'Gentoo')]
gentoo_dream = birds[(birds['island'] == 'Dream') & (birds['species'] == 'Gentoo')]
gentoo_biscoe = birds[(birds['island'] == 'Biscoe') & (birds['species'] == 'Gentoo')]
chinstrap_torgersen = birds[(birds['island'] == 'Torgersen') & (birds['species'] == 'Chinstrap')]
chinstrap_dream = birds[(birds['island'] == 'Dream') & (birds['species'] == 'Chinstrap')]
chinstrap_biscoe = birds[(birds['island'] == 'Biscoe') & (birds['species'] == 'Chinstrap')]

In [55]:
chinstrap_dream

rowid    species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
276   277  Chinstrap  Dream            46.5           17.9              192.0   
277   278  Chinstrap  Dream            50.0           19.5              196.0   
278   279  Chinstrap  Dream            51.3           19.2              193.0   
279   280  Chinstrap  Dream            45.4           18.7              188.0   
280   281  Chinstrap  Dream            52.7           19.8              197.0   
..    ...        ...    ...             ...            ...                ...   
339   340  Chinstrap  Dream            55.8           19.8              207.0   
340   341  Chinstrap  Dream            43.5           18.1              202.0   
341   342  Chinstrap  Dream            49.6           18.2              193.0   
342   343  Chinstrap  Dream            50.8           19.0              210.0   
343   344  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  weight  
276       3500.0  female  2007     3.5  
277       3900.0    male  2007     3.9  
278       3650.0    male  2007     3.6  
279       3525.0  female  2007     3.5  
280       3725.0    male  2007     3.7  
..           ...     ...   ...     ...  
339       4000.0    male  2009     4.0  
340       3400.0  female  2009     3.4  
341       3775.0    male  2009     3.8  
342       4100.0    male  2009     4.1  
343       3775.0  female  2009     3.8  

[68 rows x 10 columns]

<h1 style='font-weight:700; font-size:40px;'>Insights No. 1-3</h1>
<ul>
    <li style='list-style-type:square; color:orange;'>Chinstraps are the most common  found in Dream Island</li>
    <li style='list-style-type:square;'>Adelies are found in all of three islands, having 131 penguins recorded in Torgersen Island, 151 penguins recorded in Dream Island, and 115 penguins are recorded in Biscoe Island </li>
    <li style='list-style-type:square;'>Gentoos and found exclusively in Biscoe Island with 275 penguins recorded</li>
</ul>

<h1 style='font-weight:700; font-size:40px;'>Insights No. 4-6</h1>
<h2>Average weight per specie</h2>
<ul>
    <li style='list-style-type:square;'>Average weight of Adelies are 3.7 kilograms</li>
    <li style='list-style-type:square; color:orange;'>Average weight of Gentoos are 5.1 kilograms</li>
    <li style='list-style-type:square;'>Average weight of Chinstraps are 3.7 kilograms</li>
</ul>

In [56]:
birds['weight'] = np.round(birds['body_mass_g'] / 1000, 2)

In [57]:
birds['weight']

0      3.75
1      3.80
2      3.25
3      4.20
4      3.45
       ... 
339    4.00
340    3.40
341    3.78
342    4.10
343    3.78
Name: weight, Length: 344, dtype: float64

<h1 style='color:yellow;'>Adelies average weight across islands</h1>

In [58]:
adelie_torgersen_w = np.round(adelie_torgersen['weight'].mean(), 1)
adelie_torgersen_w

np.float64(3.7)

In [59]:
adelie_dream_w = np.round(adelie_dream['weight'].mean(), 1)
adelie_dream_w

np.float64(3.7)

In [60]:
adelie_biscoe_w = np.round(adelie_biscoe['weight'].mean(), 1)
adelie_biscoe_w

np.float64(3.7)

<h1 style='color:yellow;'>Gentoos average weight across islands</h1>

In [61]:
gentoo_torgersen_w = np.round(gentoo_torgersen['weight'].mean(), 1)
gentoo_torgersen_w

np.float64(nan)

In [62]:
gentoo_dream_w = np.round(gentoo_dream['weight'].mean(), 1)
gentoo_dream_w

np.float64(nan)

In [63]:
gentoo_biscoe_w = np.round(gentoo_biscoe['weight'].mean(), 1)
gentoo_biscoe_w

np.float64(5.1)

<h1 style='color:yellow;'>Chinstraps average weight across islands</h1>

In [64]:
chinstrap_torgersen_w = np.round(chinstrap_torgersen['weight'].mean(), 1)
chinstrap_torgersen_w

np.float64(nan)

In [65]:
chinstrap_dream_w = np.round(chinstrap_dream['weight'].mean(), 1)
chinstrap_dream_w

np.float64(3.7)

In [66]:
chinstrap_biscoe_w = np.round(chinstrap_biscoe['weight'].mean(), 1)
chinstrap_biscoe_w

np.float64(nan)

<h1 style='font-weight:700; font-size:40px;'>Insight No.7</h1>
<ul>
    <li style='list-style-type:square;'>Gentoos are the most heaviest specie averaging 5.1 kilograms</li>
    <li style='list-style-type:square;'>Adelies and Chinstraps are in second both averaging 3.7 kilograms</li>
</ul>

<h1>Insight No. x</h1>
<h2>Does sex have an impact on body mass</h2>

In [67]:
birds_agg = birds.groupby(['species']).count()
birds_agg

C:\Users\LAB4-StudentPC38\AppData\Local\Temp\ipykernel_1648\3423571731.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  birds_agg = birds.groupby(['species']).count()


rowid  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
species                                                                      
Adelie       152     152             152            152                152   
Chinstrap     68      68              68             68                 68   
Gentoo       124     124             124            124                124   

           body_mass_g  sex  year  weight  
species                                    
Adelie             152  152   152     152  
Chinstrap           68   68    68      68  
Gentoo             124  124   124     124

In [68]:
birds_agg2 = birds.groupby(['species', 'island'])['weight'].mean()

C:\Users\LAB4-StudentPC38\AppData\Local\Temp\ipykernel_1648\1384156861.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  birds_agg2 = birds.groupby(['species', 'island'])['weight'].mean()


In [69]:
birds_agg2

species    island   
Adelie     Biscoe       3.709773
           Dream        3.688393
           Torgersen    3.715962
Chinstrap  Biscoe            NaN
           Dream        3.733235
           Torgersen         NaN
Gentoo     Biscoe       5.068871
           Dream             NaN
           Torgersen         NaN
Name: weight, dtype: float64

<h1 style='font-weight:700; font-size:40px;'>Insight No. 8</h1>
<ul>
    <li style='list-style-type:square;'>2007 is the year with the highest recorded male adelie</li>
</ul>

In [70]:
birds_agg3 = birds.groupby(['species', 'island', 'year', 'sex']).size()

C:\Users\LAB4-StudentPC38\AppData\Local\Temp\ipykernel_1648\2729443699.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  birds_agg3 = birds.groupby(['species', 'island', 'year', 'sex']).size()


In [71]:
birds_agg3.unstack()

sex                       female  male
species   island    year              
Adelie    Biscoe    2007       5     5
                    2008       9     9
                    2009       8     8
          Dream     2007       9    11
                    2008       8     8
                    2009      10    10
          Torgersen 2007       8    12
                    2008       8     8
                    2009       8     8
Chinstrap Biscoe    2007       0     0
                    2008       0     0
                    2009       0     0
          Dream     2007      13    13
                    2008       9     9
                    2009      12    12
          Torgersen 2007       0     0
                    2008       0     0
                    2009       0     0
Gentoo    Biscoe    2007      16    18
                    2008      22    24
                    2009      20    24
          Dream     2007       0     0
                    2008       0     0
                    2009       0     0
          Torgersen 2007       0     0
                    2008       0     0
                    2009       0     0

<h1 style='font-weight:700; font-size:40px;'>Insights No. 9-11</h1>
<ul>
    <li style='list-style-type:square;'>Adelies weight are averaging at 3.4 kilograms for female and <span style='color:orange;'>4.0 for males</span></li>
    <li style='list-style-type:square;'>Chinstraps weight are averaging at 3.5 for female and <span style='color:orange;'>3.9 for males</span></li>
    <li style='list-style-type:square;'>Gentoos weight are averaging at 4.7 for female and <span style='color:orange;'>5.4 for males</span></li>
</ul>
<h2>Conclusion:</h2>
<ul>
    <li style='list-style-type:square;'>Male penguins are much heavier than female penguins</li>
</ul>

In [72]:
birds_agg4 = np.round(birds.groupby(['species', 'sex'])['weight'].mean(), 1)

C:\Users\LAB4-StudentPC38\AppData\Local\Temp\ipykernel_1648\1022269350.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  birds_agg4 = np.round(birds.groupby(['species', 'sex'])['weight'].mean(), 1)


In [73]:
birds_agg4.unstack()

sex        female  male
species                
Adelie        3.4   4.0
Chinstrap     3.5   3.9
Gentoo        4.7   5.4

<h1 style='font-weight:700; font-size:40px;'>Insights No. 12-15</h1>
<ul>
    <li style='list-style-type:square;'>
        There is a <span style='color:orange;'>correlation</span> between flipper length and body mass
    </li>
    <li style='list-style-type:square;'>
        There is a <span style='color:orange;'>correlation</span> between bill length and body mass
    </li>
    <li style='list-style-type:square;'>
        There is a <span style='color:orange;'>correlation</span> between bill length and flipper length
    </li>
</ul>

In [74]:
bm_fl_corr = birds['flipper_length_mm'].corr(birds['body_mass_g'])
bm_fl_corr
# weak correlation between flipper length and body mass

np.float64(0.8712017673060114)

In [75]:
bm_bd_corr = birds['bill_depth_mm'].corr(birds['body_mass_g'])
bm_bd_corr
# no correlation between bill depth and body mass

np.float64(-0.4719156211860664)

In [76]:
bm_bl_corr = birds['bill_length_mm'].corr(birds['body_mass_g'])
bm_bl_corr
# weak correlation between flipper length and body mass

np.float64(0.5951098244376304)

In [77]:
fl_bd_corr = birds['bill_depth_mm'].corr(birds['flipper_length_mm'])
fl_bd_corr
# no correlation between bill depth and flipper length

np.float64(-0.5838512164654122)

In [78]:
fl_bl_corr = birds['bill_length_mm'].corr(birds['flipper_length_mm'])
fl_bl_corr
# weak correlation between bill depth and flipper length

np.float64(0.6561813407464278)

In [79]:
bl_bd_corr = birds['bill_depth_mm'].corr(birds['bill_length_mm'])
bl_bd_corr
# no correlation between bill depth and bill length

np.float64(-0.23505287035553285)

<h1 style='font-weight:700; font-size:40px;'>Library for one way ANOVA</h1>

In [80]:
pip install scipy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [81]:
from scipy.stats import f_oneway

<h2 style='color:yellow;'>Requirements for one way ANOVA:</h2>
<ol>
    <li>Filter data and its data</li>
</ol>

In [83]:
birds[birds['sex'] == 'male']['body_mass_g']

0      3750.000000
3      4201.754386
5      3650.000000
7      4675.000000
8      3475.000000
          ...     
334    3800.000000
336    3950.000000
339    4000.000000
341    3775.000000
342    4100.000000
Name: body_mass_g, Length: 179, dtype: float64

In [84]:
birds[birds['sex'] == 'female']['body_mass_g']

1      3800.0
2      3250.0
4      3450.0
6      3625.0
12     3200.0
        ...  
335    3525.0
337    3650.0
338    3650.0
340    3400.0
343    3775.0
Name: body_mass_g, Length: 165, dtype: float64

In [90]:
f_stats, p_value = f_oneway(birds[birds['sex'] == 'male']['body_mass_g'], 
         birds[birds['sex'] == 'female']['body_mass_g'])
# The variables should be two when using f_oneway because it returns 2 values, f_stats and p_value
print(p_value)

3.0226600487050246e-15


<h1 style='font-weight:700; font-size:40px;'>Insight No. 16</h1>
<h2 style='color:yellow;'>There is a strong correlation between sex and body weight</h2>